### DISCUSSION TITLE
Best practices around testing web API controllers

### DISCUSSION THREAD

**Discussion by Bartleby2718** [nunit/nunit#4825] [General]
Noob asking!

I have a .NET web application that's tested through NUnit. [NUnit1032 from NUnit.Analyzers](https://docs.nunit.org/articles/nunit-analyzers/NUnit1032.html) says:
> An IDisposable field/property should be Disposed in a TearDown method

Does this also apply to [the `Controller` class in the `Microsoft.AspNetCore.Mvc` namespace](https://source.dot.net/#Microsoft.AspNetCore.Mvc.ViewFeatures/Controller.cs,17)? That is, if I initialize a controller in my `SetUp` method, should I call `.Dispose()` against it in the corresponding `TearDown`? Is it a bad practice to rely on garbage collection for this?

**Comment by stevenaw** [comment]
Hi @Bartleby2718 that' right, it's always a good practice to dispose a resource when you're done with it. The specifics of this within your tests may depend on how you've set them up, but as a general rule they should be disposed of within the same _scope_. In other words:
- Dispose of it in your test method if you create it within your test method
- Dispose of it via `[TearDown]` if you create it in a `[SetUp]` and store it in a class variable
- Dispose of it via `[OneTimeTearDown]` if you create it in a `[OneTimeSetUp]` and store it in a class variable

Hope this helps! :)

**Comment by stevenaw** [comment]
One other tip I could give more specific to your case. If you haven't already you may want to consider looking at the `WebAPplicationFactory` class. It resides in the ` Microsoft.AspNetCore.Mvc.Testing` NuGet package and is designed to help with creating controllers for testing purposes (https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-8.0)
